### Nama: Azmi Alfatih Shalahuddin
### NIM : 16520001
### Instansi : Institut Teknologi Bandung
### Pengerjaan Tugas : 10/ 10/ 2020
### Sumber Data Set : TakeMeOut

In [ ]:
#mengmport library
import pandas as pd
from sklearn.neighbors import NearestNeighbors
import pickle





In [ ]:
#membaca data takemeout.csv
tb = pd.read_csv("takemeout.csv")



In [ ]:
tb.head()

Timestamp  ... Seberapa penting bantuan dari calon pacar untuk kamu?
0  2020/10/31 3:39:25 PM GMT+7  ...                                                  3   
1  2020/10/31 3:39:36 PM GMT+7  ...                                                  2   
2  2020/10/31 3:39:38 PM GMT+7  ...                                                  4   
3  2020/10/31 3:39:42 PM GMT+7  ...                                                  3   
4  2020/10/31 3:39:43 PM GMT+7  ...                                                  4   

[5 rows x 8 columns]

Tabel data mentah yang belum dibersihkan dan disederhanakan

### **Data Observation**
Data yang diperoleh berisi tanggal dan waktu pengisian data, nama, jenis kelamin pengisi data, dan ukuran ukuran love language, seperti seberapa penting quality time, physical touch, word of affirmation, gift,  dan bantuan.
Judul kolom masih berupa kalimat pertanyaan yang sangat panjang, data masih belum bersih, dan belum disederhanakan. penyederhanaan data juga dilakukan dengan menyingkat judul kolom menjadi satu kata.

In [ ]:
#Membersihkan data, menghilangkan waktu input
tb = tb.iloc[:,1:] 

### **Data Cleaning**

Data yang diperoleh sudah cukup bersih. Namun, masih ada aspek yang bisa dibersihkan sebelum data diproses, seperti waktu pengisian data, karena itu bukan skor yang menunjukkan kecocokan pasangan.



In [ ]:
#menampilkan tabel
tb.head()

Siapa nama kamu?  ... Seberapa penting bantuan dari calon pacar untuk kamu?
0  A**************  ...                                                  3   
1            L****  ...                                                  2   
2     Y***********  ...                                                  4   
3             a***  ...                                                  3   
4            B****  ...                                                  4   

[5 rows x 7 columns]

In [ ]:
#Menyederhanakan nama kolom
tb.columns= ['Name', 'Gender', 'Time', 'Touch', 'Word', 'Gift', 'Help']

In [ ]:
#Menampilkan Tabel
tb.head()

Name Gender  Time  Touch  Word  Gift  Help
0  A**************  Cowok     5      5     4     1     3
1            L****  Cewek     5      5     3     2     2
2     Y***********  Cowok     4      3     4     4     4
3             a***  Cowok     5      5     5     2     3
4            B****  Cowok     5      5     5     2     4

Menampilkan contoh tabel awal yang sudah dibersihkan dan disederhanakan  sebagai gambaran.

In [ ]:
tb.describe()

Time       Touch        Word        Gift        Help
count  101.000000  101.000000  101.000000  101.000000  101.000000
mean     4.099010    3.168317    3.722772    2.732673    3.752475
std      1.212476    1.334686    1.209287    1.325829    1.143730
min      1.000000    1.000000    1.000000    1.000000    1.000000
25%      4.000000    2.000000    3.000000    2.000000    3.000000
50%      5.000000    3.000000    4.000000    3.000000    4.000000
75%      5.000000    4.000000    5.000000    4.000000    5.000000
max      5.000000    5.000000    5.000000    5.000000    5.000000

### Diperoleh **statistik kasar** dari data, yaitu:
<br/> Min = 1 untuk semua kolom
<br/> Max = 5 untuk semua kolom
<br/>Median = 5, 3, 4, 3, 4
<br/> Rerata = 4.099, 3.168317, 3.722772, 2.732673, 3.752475

Dari data, dapat dilihat bahwa rata-rata skor terbesar terdapat pada Time (Seberapa penting Quality Time bagi kamu) (4.09), ini menunjukkan bahwa kebanyakan orang-orang memprioritaskan seringnya waktu berkualitas bersama pasangan.
<br/> Bisa dilihat juga bahwa rata-rata skor paling kecil ialah Gift (Seberapa penting mendapatkan kado dari pasangan) (2.732673), menunjukkan bahwa kebanyakan orang paling mentoleransi (atau tidak mengedepankan) hadiah dari pasangannya.

### **Analisis Data**
Data telah dibersihkan dan disederhanakan menjadi seperti di atas. Sehingga, data hanya berisi hal-hal yang diperlukan dalam mencapai tujuan kita yaitu mencari calon pasangan yang cocok. terdapat 5 aspek love language yang kita miliki, yang tiap data memiliki skornya masing-masing. dapat diprediksi, semakin dekat skor love language dari lawan jenis, akan semakin cocok jika mereka berpasangan. oleh karena itu, digunakan metode pencarian skor yang terdekat untuk mencapai tujuan kita.

In [ ]:
#Memisahkan data gender cowok dan cewek untuk membentuk pasangannya
m_tb = tb[tb['Gender'] == 'Cowok']
f_tb = tb[tb['Gender'] ==  'Cewek']

In [ ]:
m_X = m_tb.drop(['Name', 'Gender'], axis =1)
f_X = f_tb.drop(['Name', 'Gender'], axis =1)

# Model Development

In [ ]:
#membuat variabel cocok sebagai 7 calon pasangan yang cocok
#Program menggunakan library Nearest Neighbor 
m_cocok = NearestNeighbors(n_neighbors=7).fit(m_X)
f_cocok = NearestNeighbors(n_neighbors=7).fit(f_X)

In [ ]:
X = tb.drop(['Name', 'Gender'], axis=1)

In [ ]:
#Melakukan uji coba pada suatu nama
test = X[tb['Name'] == 'F***'].values

In [ ]:
#membuat variabel idx sebagai pasangan yang cocok dari nama yang dimasukkan
_, idx= f_cocok.kneighbors(test)

In [ ]:
#contoh keluaran index dari pasangan yang cocok dari "F***"
idx

array([[ 5, 16,  4,  2, 15, 11,  6],
       [ 4, 16,  5,  8,  2, 15, 11]])

In [ ]:
#Mengeluarkan nama dari pasangan yg cocok
f_tb.iloc[idx[0]]['Name']

27             v****
73              R***
25              E***
20    F*************
72            N*****
62                 S
33              H***
Name: Name, dtype: object

Data yang dihasilkan berupa 7 nama dengan kecocokan terdekat, yang didapat dengan menggunakan module Nearest Neigbors

Kita mengira-ngira calon pasangan yang cocok dari masing masing skor aspek "Love Language" yang dimasukkan setiap orang dalam survey. Dengan asumsi jika "Love Language" sama atau dekat maka orang itu akan cocok, kita dapat mencari calon pasangan yang akan cocok.

# Penyimpanan Model

In [ ]:
#jika ingin mendeploy model, kita harus menyimpan model yg kita buat
pickle.dump(m_cocok, open('/content/m_model.pickle', 'wb')) 
pickle.dump(f_cocok, open('/content/f_model.pickle', 'wb')) 

In [ ]:
#agar deploy yang kita buat memakai nama yang belum dimasukkan, data dari data awal
tb.to_csv('/content/data.csv', index = False)